File: app/src/main/java/com/example/shiftly/data/parser/SmartParser.kt
package com.example.shiftly.data.parser

import com.example.shiftly.data.model.Shift
import java.util.Calendar
import java.util.TimeZone

object SmartParser {

    private const val DEFAULT_YEAR = 2026

    /**
     * Regex matching patterns like:
     * 07.09.2026 - 09:00 - 17:00 45 דקות + 50
     * 07.09 - 22:00 - 06:00 ללא + 20.5
     * 15/09/2026 10:00 - 18:30 ללא הפסקה + 30
     */
    private val lineRegex = Regex(
        """(?x)
        # Date: DD.MM or DD.MM.YYYY (or with /)
        (?<day>\d{1,2})[./](?<month>\d{1,2})(?:[./](?<year>\d{2,4}))?
        \s*[-–—:\s]\s*
        # Start Time: HH:mm
        (?<startH>\d{1,2}):(?<startM>\d{2})
        \s*[-–—:\s]\s*
        # End Time: HH:mm
        (?<endH>\d{1,2}):(?<endM>\d{2})
        # Remainder for break & tips
        (?<remainder>.*)
        """.trimIndent()
    )

    private val breakRegex = Regex("""(\d+)\s*(?:דקות|דק['״]|min|minutes)?""", RegexOption.IGNORE_CASE)
    private val noBreakRegex = Regex("""(?:ללא(?:\s+הפסקה)?|none|\b0\b|בלי)""", RegexOption.IGNORE_CASE)
    private val tipRegex = Regex("""(?:\+|\bטיפ(?:ים)?\b)\s*([0-9]+(?:\.[0-9]+)?)""", RegexOption.IGNORE_CASE)

    fun parseLine(rawLine: String, defaultHourlyRate: Double = 45.0): Shift? {
        val trimmed = rawLine.trim()
        if (trimmed.isEmpty()) return null

        val match = lineRegex.find(trimmed) ?: return null

        val day = match.groups["day"]?.value?.toIntOrNull() ?: return null
        val month = match.groups["month"]?.value?.toIntOrNull() ?: return null
        var year = match.groups["year"]?.value?.toIntOrNull() ?: DEFAULT_YEAR
        if (year < 100) {
            year += 2000
        }

        val startH = match.groups["startH"]?.value?.toIntOrNull() ?: return null
        val startM = match.groups["startM"]?.value?.toIntOrNull() ?: return null
        val endH = match.groups["endH"]?.value?.toIntOrNull() ?: return null
        val endM = match.groups["endM"]?.value?.toIntOrNull() ?: return null

        val tz = TimeZone.getDefault()
        val startCal = Calendar.getInstance(tz).apply {
            set(Calendar.YEAR, year)
            set(Calendar.MONTH, month - 1)
            set(Calendar.DAY_OF_MONTH, day)
            set(Calendar.HOUR_OF_DAY, startH)
            set(Calendar.MINUTE, startM)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }

        val endCal = Calendar.getInstance(tz).apply {
            set(Calendar.YEAR, year)
            set(Calendar.MONTH, month - 1)
            set(Calendar.DAY_OF_MONTH, day)
            set(Calendar.HOUR_OF_DAY, endH)
            set(Calendar.MINUTE, endM)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }

        // Night shift wrap-around: if end is before or equal to start, advance end date by 1 day
        if (endCal.timeInMillis <= startCal.timeInMillis) {
            endCal.add(Calendar.DAY_OF_MONTH, 1)
        }

        val remainder = match.groups["remainder"]?.value.orEmpty().trim()
        var breakMinutes = 0
        var tips = 0.0

        if (remainder.isNotEmpty()) {
            // Check for tips first
            val tipMatch = tipRegex.find(remainder)
            if (tipMatch != null) {
                tips = tipMatch.groupValues[1].toDoubleOrNull() ?: 0.0
            }

            // Check break
            if (noBreakRegex.containsMatchIn(remainder)) {
                breakMinutes = 0
            } else {
                val breakMatch = breakRegex.find(remainder)
                if (breakMatch != null) {
                    breakMinutes = breakMatch.groupValues[1].toIntOrNull() ?: 0
                }
            }
        }

        return Shift(
            startTime = startCal.timeInMillis,
            endTime = endCal.timeInMillis,
            breakDurationMinutes = breakMinutes,
            isBreakPaid = false,
            tips = tips,
            hourlyRate = defaultHourlyRate,
            jobRole = "משמרת",
            notes = "יובא אוטומטית מ-WhatsApp"
        )
    }

    fun parseMultiple(text: String, defaultHourlyRate: Double = 45.0): List<Shift> {
        val results = mutableListOf<Shift>()
        val lines = text.split("\n", ";", "\r\n")
        for (line in lines) {
            val parsed = parseLine(line, defaultHourlyRate)
            if (parsed != null) {
                results.add(parsed)
            }
        }
        return results
    }
}

File: lib/services/smart_parser.dart
import '../models/shift.dart';

class SmartParser {
  static const int defaultYear = 2026;

  static final RegExp _lineRegex = RegExp(
    r'(?<day>\d{1,2})[./](?<month>\d{1,2})(?:[./](?<year>\d{2,4}))?'
    r'\s*[-–—:\s]\s*'
    r'(?<startH>\d{1,2}):(?<startM>\d{2})'
    r'\s*[-–—:\s]\s*'
    r'(?<endH>\d{1,2}):(?<endM>\d{2})'
    r'(?<remainder>.*)',
  );

  static final RegExp _breakRegex = RegExp(
    r'(\d+)\s*(?:דקות|דק[\'״]|min|minutes)?',
    caseSensitive: false,
  );

  static final RegExp _noBreakRegex = RegExp(
    r'(?:ללא(?:\s+הפסקה)?|none|\b0\b|בלי)',
    caseSensitive: false,
  );

  static final RegExp _tipRegex = RegExp(
    r'(?:\+|\bטיפ(?:ים)?\b)\s*([0-9]+(?:\.[0-9]+)?)',
    caseSensitive: false,
  );

  /// Parse a single line of freeform shift log.
  static Shift? parseLine(String rawLine, {double defaultHourlyRate = 45.0}) {
    final trimmed = rawLine.trim();
    if (trimmed.isEmpty) return null;

    final match = _lineRegex.firstMatch(trimmed);
    if (match == null) return null;

    final day = int.tryParse(match.namedGroup('day') ?? '');
    final month = int.tryParse(match.namedGroup('month') ?? '');
    var year = int.tryParse(match.namedGroup('year') ?? '') ?? defaultYear;
    if (year < 100) {
      year += 2000;
    }

    final startH = int.tryParse(match.namedGroup('startH') ?? '');
    final startM = int.tryParse(match.namedGroup('startM') ?? '');
    final endH = int.tryParse(match.namedGroup('endH') ?? '');
    final endM = int.tryParse(match.namedGroup('endM') ?? '');

    if (day == null || month == null || startH == null || startM == null || endH == null || endM == null) {
      return null;
    }

    final startDateTime = DateTime(year, month, day, startH, startM);
    var endDateTime = DateTime(year, month, day, endH, endM);

    // Cross-midnight / night shifts: e.g., 22:00 - 06:00 wraps to the next calendar day
    if (endDateTime.isBefore(startDateTime) || endDateTime.isAtSameMomentAs(startDateTime)) {
      endDateTime = endDateTime.add(const Duration(days: 1));
    }

    final remainder = (match.namedGroup('remainder') ?? '').trim();
    int breakMinutes = 0;
    double tips = 0.0;

    if (remainder.isNotEmpty) {
      final tipMatch = _tipRegex.firstMatch(remainder);
      if (tipMatch != null) {
        tips = double.tryParse(tipMatch.group(1) ?? '0') ?? 0.0;
      }

      if (_noBreakRegex.hasMatch(remainder)) {
        breakMinutes = 0;
      } else {
        final breakMatch = _breakRegex.firstMatch(remainder);
        if (breakMatch != null) {
          breakMinutes = int.tryParse(breakMatch.group(1) ?? '0') ?? 0;
        }
      }
    }

    return Shift(
      startTime: startDateTime,
      endTime: endDateTime,
      breakDurationMinutes: breakMinutes,
      isBreakPaid: false,
      tips: tips,
      hourlyRate: defaultHourlyRate,
      jobRole: 'משמרת',
      notes: 'יובא אוטומטית מ-WhatsApp',
    );
  }

  /// Parse multiple lines (e.g. from WhatsApp messages).
  static List<Shift> parseMultiple(String text, {double defaultHourlyRate = 45.0}) {
    final results = <Shift>[];
    final lines = text.split(RegExp(r'[\r\n;]+'));
    for (final line in lines) {
      final shift = parseLine(line, defaultHourlyRate: defaultHourlyRate);
      if (shift != null) {
        results.add(shift);
      }
    }
    return results;
  }
}
